envoy/common/io/io_uring.h

```diff
@@ -0,0 +1,446 @@
+  virtual IoUringResult prepareCancel(void* cancelling_user_data, void* user_data) PURE;
+
+  /**
+   * Prepares a shutdown operation and puts it into the submission queue.
+   * Returns IoUringResult::Failed in case the submission queue is full already
+   * and IoUringResult::Ok otherwise.
+   */
+  virtual IoUringResult prepareShutdown(os_fd_t fd, int how, void* user_data) PURE;
+
+  /**
+   * Submits the entries in the submission queue to the kernel using the
+   * `io_uring_enter()` system call.
+   * Returns IoUringResult::Ok in case of success and may return
+   * IoUringResult::Busy if we over commit the number of requests. In the latter
+   * case the application should drain the completion queue by handling some completions
+   * with the forEveryCompletion() method and try again.
+   */
+  virtual IoUringResult submit() PURE;
+
+  /**
+   * Inject a request completion into the io_uring.
+   */
+  virtual void injectCompletion(os_fd_t fd, void* user_data, int32_t result) PURE;
+
+  /**
+   * Remove the injected completion for specific fd.
+   */
+  virtual void removeInjectedCompletion(os_fd_t fd,
+                                        InjectedCompletionUserDataReleasor releasor) PURE;
+};
+
+using IoUringPtr = std::unique_ptr<IoUring>;
+
+/**
+ * io_uring request type.
+ */
+struct RequestType {
+  static constexpr uint32_t Accept = 0x1;
+  static constexpr uint32_t Connect = 0x2;
+  static constexpr uint32_t Read = 0x4;
+  static constexpr uint32_t Write = 0x8;
+  static constexpr uint32_t Close = 0x10;
+  static constexpr uint32_t Cancel = 0x20;
+  static constexpr uint32_t Shutdown = 0x40;
+};
+
+class IoUringSocket;
+
+/**
+ * Abstract for io_uring I/O Request.
+ */
+class Request {
+public:
+  virtual ~Request() = default;
+
+  /**
+   * Return the request type.
+   */
+  virtual uint32_t type() const PURE;
+
+  /**
+   * Return on which io_uring socket the request belongs to.
+   */
+  virtual IoUringSocket& socket() const PURE;
+};
+
+/**
+ * The Status of IoUringSocket.
+ */
+enum IoUringSocketStatus {
+  Initialized,
+  Enabled,
+  Disabled,
+  RemoteClosed,
+  Closed,
+};
+
+class IoUringWorker;
+
+using IoUringSocketOnClosedCb = std::function<void()>;
+
+struct AcceptedSocketParam {
+  os_fd_t fd_;
+  sockaddr_storage* remote_addr_;
+  socklen_t remote_addr_len_;
+};
+
+struct ReadParam {
+  Buffer::Instance& buf_;
+  int32_t result_;
+};
+
+struct WriteParam {
+  int32_t result_;
+};
+
+/**
+ * Abstract for each socket.
+ */
+class IoUringSocket {
+public:
+  virtual ~IoUringSocket() = default;
+
+  /**
+   * Get the IoUringWorker this socket bind to.
+   */
+  virtual IoUringWorker& getIoUringWorker() const PURE;
+
+  /**
+   * Return the raw fd.
+   */
+  virtual os_fd_t fd() const PURE;
+
+  /**
+   * Close the socket.
+   * param keep_fd_open is indicated the file descriptor of the socket will be closed or not in the
+   * end. The value of `true` is used for destroy the IoUringSocket but keep the file descriptor
```

Comment:
What is the use case for deleting the socket object but keeping the descriptor open?

Reply:
This is used for migrate the socket between thread https://github.com/envoyproxy/envoy/pull/24958/commits/fed4c7547ed27f6be4d4cec744feee21f5fe43e4

For the connection balance, we will move a socket from one thread to another. For this case, we need to close IoUringSocket in the current thread, and create an new one in the new thread, but we need to keep the fd open.

I will update more description here.